command/agent/check.go

```diff
@@ -532,6 +545,33 @@ func (c *CheckTCP) check() {
 	c.Notify.UpdateCheck(c.CheckID, structs.HealthPassing, fmt.Sprintf("TCP connect %s: Success", c.TCP))
 }
 
+// checkUDP is invoked periodically to perform the UDP check
+func (c *CheckTCP) checkUDP() {
+	conn, err := c.dialer.Dial(`udp`, c.UDP)
+	if err == nil {
+		b := []byte("CONNECTED-MODE SOCKET")
+		err = conn.SetDeadline(time.Now().Add(c.dialer.Timeout))
+		if err == nil {
+			_, err = conn.Write(b)
+			if err == nil {
+				buf := make([]byte, 1024)
+				_, err = conn.Read(buf)
+				if nErr, ok := err.(net.Error); ok && nErr.Timeout() {
+					err = nil
+				}
+			}
+		}
+	}
+	if err != nil {
+		c.Logger.Printf("[WARN] agent: socket connection failed '%s': %s", c.UDP, err)
+		c.Notify.UpdateCheck(c.CheckID, structs.HealthCritical, err.Error())
+		return
```

Comment:
Is this a place we might consider using defer? In the event of an error the function returns and the connection is never closed? We might need to move the `c.Notify` call but the debug logger is always desired?